pkg/flowaggregator/s3uploader/s3uploader_test.go

```diff
@@ -0,0 +1,217 @@
+		nameRand:         nameRand,
+	}
+	testRecord := flowrecordtesting.PrepareTestFlowRecord()
+	s3UploadProc.writeRecordToBuffer(testRecord)
+	s3UploadProc.cachedRecordCount = 1
+	err := s3UploadProc.batchUploadAll(context.TODO())
+	assert.NoError(t, err)
+	assert.Equal(t, 0, len(s3UploadProc.bufferQueue))
+	assert.Equal(t, 0, len(s3UploadProc.buffersToUpload))
+	assert.Equal(t, "", s3UploadProc.currentBuffer.String())
+	assert.Equal(t, 0, s3UploadProc.cachedRecordCount)
+	assert.Contains(t, mockS3Uploader.testReader.String(), recordStrIPv4)
+}
+
+func TestBatchUploadAllError(t *testing.T) {
+	s3uploader := S3Uploader{}
+	// #nosec G404: random number generator not used for security purposes
+	nameRand := rand.New(rand.NewSource(seed))
+	s3UploadProc := S3UploadProcess{
+		bucketName:       "test-bucket-name",
+		compress:         false,
+		maxRecordPerFile: 10,
+		currentBuffer:    &bytes.Buffer{},
+		bufferQueue:      make([]*bytes.Buffer, 0),
+		buffersToUpload:  make([]*bytes.Buffer, 0, maxNumBuffersPendingUpload),
+		s3UploaderAPI:    s3uploader,
+		nameRand:         nameRand,
+	}
+	cfg, _ := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-west-2"))
+	s3UploadProc.awsS3Client = s3.NewFromConfig(cfg)
+	s3UploadProc.awsS3Uploader = s3manager.NewUploader(s3UploadProc.awsS3Client)
+
+	testRecord := flowrecordtesting.PrepareTestFlowRecord()
+	s3UploadProc.writeRecordToBuffer(testRecord)
+	s3UploadProc.cachedRecordCount = 1
+	// It is expected to fail when calling uploadFile, as the correct S3 bucket
+	// configuration is not provided.
+	err := s3UploadProc.batchUploadAll(context.TODO())
+	assert.Equal(t, 1, len(s3UploadProc.buffersToUpload))
+	assert.Equal(t, 0, len(s3UploadProc.bufferQueue))
+	assert.Equal(t, "", s3UploadProc.currentBuffer.String())
+	assert.Equal(t, 0, s3UploadProc.cachedRecordCount)
+	expectedErrMsg := "error when uploading file to S3: operation error S3: PutObject, https response error StatusCode: 301"
+	assert.Contains(t, err.Error(), expectedErrMsg)
+}
+
+func TestFlowRecordPeriodicCommit(t *testing.T) {
+	mockS3Uploader := mockS3Uploader{testReader: &bytes.Buffer{}}
+	// #nosec G404: random number generator not used for security purposes
+	nameRand := rand.New(rand.NewSource(seed))
+	s3UploadProc := S3UploadProcess{
+		compress:         false,
+		maxRecordPerFile: 10,
+		uploadInterval:   100 * time.Millisecond,
+		currentBuffer:    &bytes.Buffer{},
+		bufferQueue:      make([]*bytes.Buffer, 0),
+		buffersToUpload:  make([]*bytes.Buffer, 0, maxNumBuffersPendingUpload),
+		s3UploaderAPI:    mockS3Uploader,
+		nameRand:         nameRand,
+	}
+	testRecord := flowrecordtesting.PrepareTestFlowRecord()
+	s3UploadProc.writeRecordToBuffer(testRecord)
+	s3UploadProc.cachedRecordCount = 1
+	s3UploadProc.startExportProcess()
+	// wait for ticker to tick
+	err := wait.PollImmediate(10*time.Millisecond, 200*time.Millisecond, func() (bool, error) {
+		if mockS3Uploader.testReader.Len() != 0 {
```

Comment:
is a race condition possible here? I guess if the race detector doesn't report one when the tests are run in CI, it should be ok

Reply:
yes, it's possible. added a lock for the testReader.

I previously ran it for a few times locally and didn't encounter any issue, but the last run in github workflow reported race condition.

Reply:
did you use the `-race` flag when you ran the tests locally?

Reply:
yes, I always use the flag. Probably because I only ran it once or twice? can't remember clearly. After finding the CI test failure, I ran it again locally and it also reports race condition.